fc/language/values.py

```diff
@@ -1,6 +1,12 @@
+"""
+Classes for values in the protocol language (e.g. numbers, but also lambda
+functions and data loading).
+"""
 
+import os
 import numpy as np
 
+from ..data_loading import load2d
```

Comment:
This could cause issues as it sets up a circular import.

Reply:
What's your preferred solution? I stick the import in the method that uses it, or change it to `from .. import data_loading`
I feel the second is slightly cleaner?

Reply:
The former is the canonical solution to such problems I believe (and what was done before). Is the latter a new solution in Python 3 (I don't think it worked in Python 2 at least)?

Reply:
Hmmmm! I thought the 2nd worked because the sequence of events would be

- start importing module X
- X imports Y, so start importing Y
- Y imports X, so Y gets a pointer to the unfinished module X
- Some method in Y uses currently unknown bits of X, but Python doesn't care yet as it doesn't check the function contents until you call it
- Y finishes importing
- X gets a pointer to Y, and continues importing

So it would be ok because nothing at the top level in Y accessed members of X, just X itself
(which means there's a difference between import X and from X import Z)

Reply:
I _think_ that's how it works. Don't quote me on that!

Reply:
Some discussion here https://stackoverflow.com/questions/744373/circular-or-cyclic-imports-in-python

Reply:
Yeah, that's basically how it works, so you're safe if your modules basically just define classes & functions (and import the whole module rather than individual bits from it), but if they "do" anything during import you can get errors. I think this used to bite with the syntax parser especially because that did more stuff at import time; with the removal of the XML support we might not hit this as easily!

So actually given tests pass we could keep things as they are for now. Or do `from .. import data_loading` if you prefer to be safer.